src/conduits/RwaCounduits3.integration.t.sol

```diff
@@ -0,0 +1,377 @@
+        assertEq(dai.balanceOf(testUrn), urnDaiBalanceBefore + gemToDai(amt));
+    }
+
+    function testRevertInputConduitOnSwapAboveLine() public {
+        uint256 gemAmount = ART_LEFT + 1; // more then ART left
+
+        assertEq(gem.balanceOf(address(inputConduit)), 0);
+
+        gem.transfer(address(inputConduit), gemAmount);
+
+        assertEq(gem.balanceOf(address(inputConduit)), gemAmount);
+
+        assertEq(dai.balanceOf(testUrn), 0);
+
+        vm.expectRevert("Vat/ceiling-exceeded");
+        inputConduit.push();
+
+        assertEq(gem.balanceOf(address(inputConduit)), gemAmount);
+        assertEq(dai.balanceOf(testUrn), 0);
+    }
+
+    /*//////////////////////////////////
+               Outnput Conduit Tests
+    //////////////////////////////////*/
+
+    function testOutputConduitPush() public {
+        assertEq(outputConduit.to(), me);
+        uint256 daiAmount = URN_INK;
+
+        uint256 gemBalanceBefore = gem.balanceOf(me);
+        dai.transfer(address(outputConduit), daiAmount);
+
+        outputConduit.push();
+
+        assertEq(gem.balanceOf(address(me)), gemBalanceBefore + outputConduit.expectedGemAmt(daiAmount));
+        // We lose some dust because of decimals diif dai.decimals() > gem.decimals(), which can be exited using 'quit' method
+        assertApproxEqAbs(dai.balanceOf(address(outputConduit)), 0, GEM_DAI_DIFF_DECIMALS);
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testOutputConduitPushAmountFuzz(uint256 wadAmt) public {
+        assertEq(outputConduit.to(), me);
+
+        dai.transfer(address(outputConduit), getDaiInAmount(URN_INK / GEM_DAI_DIFF_DECIMALS));
+        uint256 cDaiBalance = dai.balanceOf(address(outputConduit));
+        uint256 gemBalance = gem.balanceOf(me);
+
+        wadAmt = bound(wadAmt, 10**18, cDaiBalance);
+
+        outputConduit.push(wadAmt);
+
+        assertEq(gem.balanceOf(me), gemBalance + outputConduit.expectedGemAmt(wadAmt));
+        // We lose some dust because of decimals diif dai.decimals() > gem.decimals(), which can be exited using 'quit' method
+        assertApproxEqAbs(
+            dai.balanceOf(address(outputConduit)),
+            cDaiBalance - getDaiInAmount(outputConduit.expectedGemAmt(wadAmt)),
+            GEM_DAI_DIFF_DECIMALS
+        );
+        assertEq(outputConduit.to(), address(0));
+    }
+
+    function testRevertOutputConduitOnInsufficientGemAmountInPsm() public {
+        uint256 daiAmount = getDaiInAmount(URN_INK / GEM_DAI_DIFF_DECIMALS + 1 * GEM_DECIMALS); // more then INK
+
+        assertEq(gem.balanceOf(address(outputConduit)), 0);
+
+        dai.transfer(address(outputConduit), daiAmount);
+
+        assertEq(dai.balanceOf(address(outputConduit)), daiAmount);
+
+        vm.expectRevert();
+        // It will revert on vat.frob()
+        // urn.ink = _add(urn.ink, dink); // _add method will revert with empty message because ink = 1000 and dink = -1100
```

Comment:
I think the comment is wrong here, we probably just have to say ink < dink

Reply:
Done